File: CT3/MPP/fine-grained-bank-smvfe/java/BankImpl.java
/**
 * Bank implementation.
 *
 * <p>:TODO: This implementation has to be made thread-safe.
 *
 * @author :TODO: LastName FirstName
 */
public class BankImpl implements Bank {
    /**
     * An array of accounts by index.
     */
    private final Account[] accounts;

    /**
     * Creates new bank instance.
     * @param n the number of accounts (numbered from 0 to n-1).
     */
    public BankImpl(int n) {
        accounts = new Account[n];
        for (int i = 0; i < n; i++) {
            accounts[i] = new Account();
        }
    }

    @Override
    public int getAccountsCount() {
        return accounts.length;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long amount(int id) {
        return accounts[id].amount;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long getTotalAmount() {
        long sum = 0;
        for (Account account : accounts) {
            sum += account.amount;
        }
        return sum;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long deposit(int id, long amount) {
        if (amount <= 0)
            throw new IllegalArgumentException("Invalid amount: " + amount);
        Account account = accounts[id];
        if (amount > MAX_AMOUNT || account.amount + amount > MAX_AMOUNT)
            throw new IllegalStateException("Overflow");
        account.amount += amount;
        return account.amount;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long withdraw(int id, long amount) {
        if (amount <= 0)
            throw new IllegalArgumentException("Invalid amount: " + amount);
        Account account = accounts[id];
        if (account.amount - amount < 0)
            throw new IllegalStateException("Underflow");
        account.amount -= amount;
        return account.amount;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public void transfer(int fromId, int toId, long amount) {
        if (amount <= 0)
            throw new IllegalArgumentException("Invalid amount: " + amount);
        if (fromId == toId)
            throw new IllegalArgumentException("fromId == toId");
        Account from = accounts[fromId];
        Account to = accounts[toId];
        if (amount > from.amount)
            throw new IllegalStateException("Underflow");
        else if (amount > MAX_AMOUNT || to.amount + amount > MAX_AMOUNT)
            throw new IllegalStateException("Overflow");
        from.amount -= amount;
        to.amount += amount;
    }

    /**
     * Private account data structure.
     */
    static class Account {
        /**
         * Amount of funds in this account.
         */
        long amount;
    }
}

File: CT3/MPP/fine-grained-bank-smvfe/src/BankImpl.kt
/**
 * Bank implementation.
 *
 * :TODO: This implementation has to be made thread-safe.
 *
 * @author Мочеков Семён
 */
class BankImpl(n: Int) : Bank {
    private val accounts: Array<Account> = Array(n) { Account() }

    override val accountsCount: Int
        get() = accounts.size

    /**
     * :TODO: This method has to be made thread-safe.
     */
    override fun amount(id: Int): Long {
        val account = accounts[id]
        account.lock.lock()

        try {
            return account.amount
        } finally {
            account.lock.unlock()
        }
    }

    override val totalAmount: Long
        get() {
            accounts.forEach { 
                it.lock.lock()
            }

            try {
                return accounts.sumOf {
                    it.amount
                }
            } finally {
                accounts.forEach {
                    it.lock.unlock()
                }
            }
        }

    /**
     * :TODO: This method has to be made thread-safe.
     */
    override fun deposit(id: Int, amount: Long): Long {
        require(amount > 0) { "Invalid amount: $amount" }
        val account = accounts[id]
        account.lock.lock()

        try {
            check(!(amount > Bank.MAX_AMOUNT || account.amount + amount > Bank.MAX_AMOUNT)) {
                "Overflow"
            }
            account.amount += amount
            return account.amount
        } finally {
            account.lock.unlock()
        }
    }

    /**
     * :TODO: This method has to be made thread-safe.
     */
    override fun withdraw(id: Int, amount: Long): Long {
        require(amount > 0) { "Invalid amount: $amount" }
        val account = accounts[id]
        account.lock.lock()
        
        try {
            check(account.amount - amount >= 0) { "Underflow" }
            account.amount -= amount
            return account.amount
        } finally {
            account.lock.unlock()
        }
    }

    /**
     * :TODO: This method has to be made thread-safe.
     */
    override fun transfer(fromId: Int, toId: Int, amount: Long) {
        require(amount > 0) {
            "Invalid amount: $amount"
        }
        require(fromId != toId) {
            "fromId == toId"
        }
        val fromAccount = accounts[fromId]
        val toAccount = accounts[toId]

        if (fromId < toId) {
            fromAccount.lock.lock()
            toAccount.lock.lock()
        } else {
            toAccount.lock.lock()
            fromAccount.lock.lock()
        }
        try {
            check(amount <= fromAccount.amount) {
                "Underflow"
            }
            check(!(amount > Bank.MAX_AMOUNT || toAccount.amount + amount > Bank.MAX_AMOUNT)) {
                "Overflow"
            }
            fromAccount.amount -= amount
            toAccount.amount += amount
        } finally {
            if (fromId < toId) {
                toAccount.lock.unlock()
                fromAccount.lock.unlock()
            } else {
                fromAccount.lock.unlock()
                toAccount.lock.unlock()
            }
        }
    }

    /**
     * Private account data structure.
     */
    class Account {
        /**
         * Amount of funds in this account.
         */
        var amount: Long = 0
        val lock = java.util.concurrent.locks.ReentrantLock()
    }
}